// @datadog/browser-rum/cjs/domain/record/serializationVerifier.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSerializationVerifier = createSerializationVerifier;
const browser_core_1 = require("@datadog/browser-core");
const types_1 = require("../../types");
const recordingScope_1 = require("./recordingScope");
const serialization_1 = require("./serialization");
function createSerializationVerifier(scope, onError) {
    // Node id numbering differs between the V1 and Change serialization
    // algorithms, so if we want to be able to compare their output, we need to replace the
    // 'input' node ids in each Change record with the 'output' node ids produced by the V1
    // serialization algorithm. This NodeIdRemapper will handle that task.
    const nodeIdRemapper = (0, serialization_1.createCopyingNodeIdRemapper)();
    scope.nodeIds = nodeIdRemapper.outputNodeIds;
    // Create a separate recording scope to use for Change serialization.
    const changeScope = (0, recordingScope_1.createRecordingScope)(scope.configuration, scope.elementsScrollPositions, {
        addShadowRoot: browser_core_1.noop,
        removeShadowRoot: browser_core_1.noop,
        flush: browser_core_1.noop,
        stop: browser_core_1.noop,
    });
    changeScope.nodeIds = nodeIdRemapper.inputNodeIds;
    // Create the converter that we'll use to transform the records generated by Change
    // serialization into the V1 format.
    let converter = (0, serialization_1.createChangeConverter)();
    const { unsubscribe } = scope.serializeObservable.subscribe((event) => {
        let changeRecord;
        const emitRecord = (record) => {
            changeRecord = record;
        };
        switch (event.type) {
            case 'full':
                // Create a new converter since we've transitioned to a new view.
                converter = (0, serialization_1.createChangeConverter)();
                (0, serialization_1.serializeFullSnapshotAsChange)(event.timestamp, event.kind, event.target, emitRecord, browser_core_1.noop, changeScope);
                break;
            case 'incremental':
                (0, serialization_1.serializeMutationsAsChange)(event.timestamp, event.target, emitRecord, browser_core_1.noop, changeScope);
                break;
            default:
                event;
                return;
        }
        if (!changeRecord) {
            onError('No change record serialized', event);
            return;
        }
        try {
            const v1Record = event.v1;
            const convertedChangeRecord = converter.convert(changeRecord, {
                nodeIdRemapper,
                timestamp: v1Record.timestamp,
            });
            verifySnapshotsMatch(v1Record, convertedChangeRecord, onError);
        }
        catch (error) {
            onError('Error verifying change record', error);
        }
    });
    return {
        stop: unsubscribe,
    };
}
function verifySnapshotsMatch(expectedRecord, actualRecord, onError) {
    if (expectedRecord.type !== types_1.RecordType.IncrementalSnapshot ||
        expectedRecord.data.source !== types_1.IncrementalSource.Mutation ||
        actualRecord.type !== types_1.RecordType.IncrementalSnapshot ||
        actualRecord.data.source !== types_1.IncrementalSource.Mutation) {
        // We expect full snapshots to be byte-for-byte equivalent, full stop.
        verifyByteForByteEquivalence('full snapshot', expectedRecord, actualRecord, onError);
        return;
    }
    // Verify that everything but the data is byte-for-byte equivalent.
    verifyByteForByteEquivalence('incremental snapshot wrapper', { ...expectedRecord, data: null }, { ...actualRecord, data: null }, onError);
    // Within the data, verify that everything but attribute mutations and node remove
    // mutations are byte-for-byte equivalent.
    verifyByteForByteEquivalence('incremental snapshot data (without attributes and removes)', { ...expectedRecord.data, attributes: null, removes: null }, { ...actualRecord.data, attributes: null, removes: null }, onError);
    // Verify that the attribute mutations are byte-for-byte equivalent except for order.
    const orderAttributeMutations = (a, b) => a.id - b.id;
    verifyByteForByteEquivalence('incremental snapshot attribute mutations', expectedRecord.data.attributes.sort(orderAttributeMutations), actualRecord.data.attributes.sort(orderAttributeMutations), onError);
    // Gather the set of removes performed by V1 serialization.
    const expectedRemoves = expectedRecord.data.removes;
    const expectedRemovedNodes = new Set();
    for (const remove of expectedRemoves) {
        expectedRemovedNodes.add(remove.id);
    }
    // Gather the set of removes performed by Change serialization. Change serialization is
    // more conservative about remove reporting, so it will report some removes that V1
    // serialization deliberately skips. To paper over that difference, we filter out
    // removes that V1 serialization didn't also report.
    const actualRemoves = actualRecord.data.removes.filter((remove) => expectedRemovedNodes.has(remove.id));
    // Verify that the removes that remain are byte-for-byte equivalent.
    verifyByteForByteEquivalence('incremental snapshot removes', expectedRemoves, actualRemoves, onError);
}
function verifyByteForByteEquivalence(context, expected, actual, onError) {
    const expectedString = JSON.stringify(expected);
    const actualString = JSON.stringify(actual);
    if (actualString !== expectedString) {
        onError('Serialization mismatch', {
            context,
            mismatch: createSerializationMismatchContext(expectedString, actualString),
        });
    }
}
function createSerializationMismatchContext(expected, actual) {
    const length = Math.min(expected.length, actual.length);
    try {
        let firstDifferenceIndex = 0;
        while (firstDifferenceIndex < length && expected[firstDifferenceIndex] === actual[firstDifferenceIndex]) {
            firstDifferenceIndex++;
        }
        return {
            expected: getStringNearPosition(expected, firstDifferenceIndex),
            actual: getStringNearPosition(actual, firstDifferenceIndex),
        };
    }
    catch (e) {
        return { firstDifferenceError: JSON.stringify(e) };
    }
}
function getStringNearPosition(str, index) {
    const leftContextStart = Math.max(index - 50, 0);
    const rightContextEnd = Math.min(index + 150, str.length);
    return `${str.substring(leftContextStart, index)}(!)${str.substring(index, rightContextEnd)}`;
}
