// @datadog/browser-rum/esm/domain/record/serialization/index.js
export { createChangeConverter, createCopyingNodeIdRemapper, createIdentityNodeIdRemapper } from './conversions';
export { isFullSnapshotChangeRecordsEnabled, isIncrementalSnapshotChangeRecordsEnabled } from './experimentalFeatures';
export { createChildInsertionCursor, createRootInsertionCursor } from './insertionCursor';
export { getElementInputValue } from './serializationUtils';
export { serializeFullSnapshot } from './serializeFullSnapshot';
export { serializeFullSnapshotAsChange } from './serializeFullSnapshotAsChange';
export { serializeMutations } from './serializeMutations';
export { serializeMutationsAsChange } from './serializeMutationsAsChange';
export { serializeNode } from './serializeNode';
export { serializeNodeAsChange } from './serializeNodeAsChange';
export { serializeAttribute } from './serializeAttribute';
export { createSerializationStats, updateSerializationStats, aggregateSerializationStats } from './serializationStats';
export { serializeChangesInTransaction, serializeInTransaction } from './serializationTransaction';
