// @datadog/browser-rum/cjs/domain/profiling/actionHistory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ACTION_ID_HISTORY_TIME_OUT_DELAY = void 0;
exports.createActionHistory = createActionHistory;
const browser_core_1 = require("@datadog/browser-core");
exports.ACTION_ID_HISTORY_TIME_OUT_DELAY = browser_core_1.SESSION_TIME_OUT_DELAY;
function createActionHistory(lifeCycle) {
    const history = (0, browser_core_1.createValueHistory)({
        expireDelay: exports.ACTION_ID_HISTORY_TIME_OUT_DELAY,
    });
    lifeCycle.subscribe(15 /* LifeCycleEventType.ACTION_STARTED */, (actionStart) => {
        history.add({
            id: actionStart.id,
            // The label is temporarily empty since we need to account for customers that might
            // redact the action name in the beforeSend callback. We will either try to do patch this
            // behavior, or remove the label field entirely.
            label: '',
            startClocks: actionStart.startClocks,
            duration: undefined,
        }, actionStart.startClocks.relative);
    });
    lifeCycle.subscribe(12 /* LifeCycleEventType.RAW_RUM_EVENT_COLLECTED */, ({ rawRumEvent, startClocks, duration }) => {
        if (rawRumEvent.type === 'action') {
            const historyEntry = history
                .getEntries(startClocks.relative)
                .find((entry) => entry.value.id === rawRumEvent.action.id);
            const durationForEntry = duration !== null && duration !== void 0 ? duration : 0;
            if (historyEntry) {
                historyEntry.value.duration = durationForEntry;
                historyEntry.close((0, browser_core_1.addDuration)(startClocks.relative, durationForEntry));
            }
            else {
                history
                    .add({
                    id: rawRumEvent.action.id,
                    // The label is temporarily empty since we need to account for customers that might
                    // redact the action name in the beforeSend callback. We will either try to do patch this
                    // behavior, or remove the label field entirely.
                    label: '',
                    startClocks,
                    duration,
                }, startClocks.relative)
                    .close((0, browser_core_1.addDuration)(startClocks.relative, durationForEntry));
            }
        }
    });
    return history;
}
