// @datadog/browser-rum/cjs/domain/profiling/vitalHistory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VITAL_ID_HISTORY_TIME_OUT_DELAY = void 0;
exports.createVitalHistory = createVitalHistory;
const browser_core_1 = require("@datadog/browser-core");
exports.VITAL_ID_HISTORY_TIME_OUT_DELAY = browser_core_1.SESSION_TIME_OUT_DELAY;
function createVitalHistory(lifeCycle) {
    const history = (0, browser_core_1.createValueHistory)({
        expireDelay: exports.VITAL_ID_HISTORY_TIME_OUT_DELAY,
    });
    lifeCycle.subscribe(16 /* LifeCycleEventType.VITAL_STARTED */, (vitalStart) => {
        history.add({
            id: vitalStart.id,
            startClocks: vitalStart.startClocks,
            duration: undefined,
            label: vitalStart.name,
        }, vitalStart.startClocks.relative);
    });
    lifeCycle.subscribe(12 /* LifeCycleEventType.RAW_RUM_EVENT_COLLECTED */, ({ rawRumEvent, startClocks, duration }) => {
        if (rawRumEvent.type === 'vital') {
            const historyEntry = history
                .getEntries(startClocks.relative)
                .find((entry) => entry.value.id === rawRumEvent.vital.id);
            if (historyEntry) {
                historyEntry.value.duration = duration;
                historyEntry.close((0, browser_core_1.addDuration)(startClocks.relative, duration));
            }
            else {
                history
                    .add({
                    id: rawRumEvent.vital.id,
                    startClocks,
                    duration,
                    label: rawRumEvent.vital.name,
                }, startClocks.relative)
                    .close((0, browser_core_1.addDuration)(startClocks.relative, duration));
            }
        }
    });
    return history;
}
