// @datadog/browser-rum/esm/domain/profiling/actionHistory.js
import { addDuration, createValueHistory, SESSION_TIME_OUT_DELAY } from '@datadog/browser-core';
export const ACTION_ID_HISTORY_TIME_OUT_DELAY = SESSION_TIME_OUT_DELAY;
export function createActionHistory(lifeCycle) {
    const history = createValueHistory({
        expireDelay: ACTION_ID_HISTORY_TIME_OUT_DELAY,
    });
    lifeCycle.subscribe(15 /* LifeCycleEventType.ACTION_STARTED */, (actionStart) => {
        history.add({
            id: actionStart.id,
            // The label is temporarily empty since we need to account for customers that might
            // redact the action name in the beforeSend callback. We will either try to do patch this
            // behavior, or remove the label field entirely.
            label: '',
            startClocks: actionStart.startClocks,
            duration: undefined,
        }, actionStart.startClocks.relative);
    });
    lifeCycle.subscribe(12 /* LifeCycleEventType.RAW_RUM_EVENT_COLLECTED */, ({ rawRumEvent, startClocks, duration }) => {
        if (rawRumEvent.type === 'action') {
            const historyEntry = history
                .getEntries(startClocks.relative)
                .find((entry) => entry.value.id === rawRumEvent.action.id);
            const durationForEntry = duration !== null && duration !== void 0 ? duration : 0;
            if (historyEntry) {
                historyEntry.value.duration = durationForEntry;
                historyEntry.close(addDuration(startClocks.relative, durationForEntry));
            }
            else {
                history
                    .add({
                    id: rawRumEvent.action.id,
                    // The label is temporarily empty since we need to account for customers that might
                    // redact the action name in the beforeSend callback. We will either try to do patch this
                    // behavior, or remove the label field entirely.
                    label: '',
                    startClocks,
                    duration,
                }, startClocks.relative)
                    .close(addDuration(startClocks.relative, durationForEntry));
            }
        }
    });
    return history;
}
