// @datadog/browser-rum/esm/domain/profiling/vitalHistory.js
import { addDuration, createValueHistory, SESSION_TIME_OUT_DELAY } from '@datadog/browser-core';
export const VITAL_ID_HISTORY_TIME_OUT_DELAY = SESSION_TIME_OUT_DELAY;
export function createVitalHistory(lifeCycle) {
    const history = createValueHistory({
        expireDelay: VITAL_ID_HISTORY_TIME_OUT_DELAY,
    });
    lifeCycle.subscribe(16 /* LifeCycleEventType.VITAL_STARTED */, (vitalStart) => {
        history.add({
            id: vitalStart.id,
            startClocks: vitalStart.startClocks,
            duration: undefined,
            label: vitalStart.name,
        }, vitalStart.startClocks.relative);
    });
    lifeCycle.subscribe(12 /* LifeCycleEventType.RAW_RUM_EVENT_COLLECTED */, ({ rawRumEvent, startClocks, duration }) => {
        if (rawRumEvent.type === 'vital') {
            const historyEntry = history
                .getEntries(startClocks.relative)
                .find((entry) => entry.value.id === rawRumEvent.vital.id);
            if (historyEntry) {
                historyEntry.value.duration = duration;
                historyEntry.close(addDuration(startClocks.relative, duration));
            }
            else {
                history
                    .add({
                    id: rawRumEvent.vital.id,
                    startClocks,
                    duration,
                    label: rawRumEvent.vital.name,
                }, startClocks.relative)
                    .close(addDuration(startClocks.relative, duration));
            }
        }
    });
    return history;
}
