// @datadog/browser-rum/esm/domain/record/serialization/conversions/mutationLog.js
export function createMutationLog() {
    const self = {
        onAttributeChanged(nodeId, name) {
            let changedAttributes = self.attributeChanges.get(nodeId);
            if (!changedAttributes) {
                changedAttributes = new Set();
                self.attributeChanges.set(nodeId, changedAttributes);
            }
            changedAttributes.add(name);
        },
        onNodeConnected(nodeId) {
            self.nodeAdds.add(nodeId);
        },
        onNodeDisconnected(nodeId, parentId) {
            self.nodeRemoves.set(nodeId, parentId);
        },
        onTextChanged(nodeId) {
            self.textChanges.add(nodeId);
        },
        clear() {
            self.attributeChanges.clear();
            self.nodeAdds.clear();
            self.nodeRemoves.clear();
            self.textChanges.clear();
        },
        attributeChanges: new Map(),
        nodeAdds: new Set(),
        nodeRemoves: new Map(),
        textChanges: new Set(),
    };
    return self;
}
