// @datadog/browser-rum/esm/domain/record/serialization/conversions/vDocument.js
import { IncrementalSource, RecordType } from '../../../../types';
import { createMutationLog } from './mutationLog';
import { createV1RenderOptions } from './renderOptions';
import { createVNode } from './vNode';
import { createVStyleSheet } from './vStyleSheet';
export function createVDocument() {
    const mutations = createMutationLog();
    let nextNodeId = 0;
    const nodesById = new Map();
    let rootId;
    let nextStyleSheetId = 0;
    const styleSheetsById = new Map();
    const self = {
        createNode(data) {
            const id = nextNodeId++;
            const node = createVNode(self, id, data);
            nodesById.set(id, node);
            return node;
        },
        getNodeById(id) {
            const node = nodesById.get(id);
            if (!node) {
                throw new Error(`Reference to unknown node: ${id}`);
            }
            return node;
        },
        createStyleSheet(data) {
            const id = nextStyleSheetId++;
            const sheet = createVStyleSheet(self, id, data);
            styleSheetsById.set(id, sheet);
            return sheet;
        },
        getStyleSheetById(id) {
            const sheet = styleSheetsById.get(id);
            if (!sheet) {
                throw new Error(`Reference to unknown stylesheet: ${id}`);
            }
            return sheet;
        },
        onAttributeChanged(node, name) {
            mutations.onAttributeChanged(node.id, name);
        },
        onNodeConnected(node, parent) {
            if (node.state !== 'new') {
                throw new Error(`Moving connected node: ${node.id}`);
            }
            if (parent === undefined) {
                if (rootId !== undefined) {
                    throw new Error('Replacing existing root node');
                }
                rootId = node.id;
            }
            node.state = 'connected';
            mutations.onNodeConnected(node.id);
        },
        onNodeDisconnected(node, parent) {
            node.forSelfAndEachDescendant((descendant) => {
                if (descendant.state === 'disconnected') {
                    throw new Error(`Disconnecting node which isn't connected: ${descendant.id}`);
                }
                descendant.state = 'disconnected';
            });
            if (!parent) {
                throw new Error('Disconnecting the root node');
            }
            mutations.onNodeDisconnected(node.id, parent.id);
        },
        onTextChanged(node) {
            mutations.onTextChanged(node.id);
        },
        get mutations() {
            return mutations;
        },
        naturalRendering() {
            if (rootId !== undefined && mutations.nodeAdds.has(rootId)) {
                return RecordType.FullSnapshot;
            }
            return RecordType.IncrementalSnapshot;
        },
        render(options) {
            if (self.naturalRendering() === RecordType.FullSnapshot) {
                return self.renderAsFullSnapshot(options);
            }
            return self.renderAsIncrementalSnapshot(options);
        },
        renderAsFullSnapshot(renderOptions = {}) {
            var _a, _b;
            const options = createV1RenderOptions(renderOptions);
            const root = self.root;
            if (!root) {
                throw new Error('No root node found');
            }
            let scrollLeft = 0;
            let scrollTop = 0;
            if (root.data.kind === '#document' || root.data.kind === '#element') {
                scrollLeft = (_a = root.data.scrollLeft) !== null && _a !== void 0 ? _a : 0;
                scrollTop = (_b = root.data.scrollTop) !== null && _b !== void 0 ? _b : 0;
            }
            return {
                data: {
                    node: root.render(options),
                    initialOffset: { left: scrollLeft, top: scrollTop },
                },
                type: RecordType.FullSnapshot,
                timestamp: options.timestamp,
            };
        },
        renderAsIncrementalSnapshot(renderOptions = {}) {
            var _a;
            const options = createV1RenderOptions(renderOptions);
            const root = self.root;
            if (!root) {
                throw new Error('No root node found');
            }
            const remappedId = (id) => options.nodeIdRemapper.remap(id);
            const addMutations = [];
            root.forEachAddedNodeRoot(mutations.nodeAdds, (node) => {
                var _a;
                if (!node.parent) {
                    throw new Error(`Can't render incremental add of root node ${node.id}`);
                }
                addMutations.push({
                    nextId: ((_a = node.nextSibling) === null || _a === void 0 ? void 0 : _a.id) !== undefined ? remappedId(node.nextSibling.id) : null,
                    parentId: remappedId(node.parent.id),
                    node: node.render(options),
                });
            });
            const removeMutations = [];
            for (const [id, parentId] of mutations.nodeRemoves) {
                removeMutations.push({ parentId: remappedId(parentId), id: remappedId(id) });
            }
            const textMutations = [];
            for (const id of mutations.textChanges) {
                const node = self.getNodeById(id);
                if (node.data.kind !== '#text') {
                    throw new Error(`Can't render incremental text mutation of ${node.data.kind} node ${id}`);
                }
                textMutations.push({ id: remappedId(id), value: node.data.textContent });
            }
            const attributeMutations = [];
            for (const [id, attrNames] of mutations.attributeChanges) {
                const node = self.getNodeById(id);
                if (node.data.kind !== '#element') {
                    throw new Error(`Can't render incremental attribute mutation of ${node.data.kind} node ${id}`);
                }
                const attributes = {};
                for (const attrName of attrNames) {
                    attributes[attrName] = (_a = node.data.attributes[attrName]) !== null && _a !== void 0 ? _a : null;
                }
                attributeMutations.push({ id: remappedId(id), attributes });
            }
            return {
                data: {
                    source: IncrementalSource.Mutation,
                    adds: addMutations,
                    removes: removeMutations,
                    texts: textMutations,
                    attributes: attributeMutations,
                },
                type: RecordType.IncrementalSnapshot,
                timestamp: options.timestamp,
            };
        },
        get root() {
            return rootId === undefined ? undefined : self.getNodeById(rootId);
        },
        set root(node) {
            self.onNodeConnected(node, undefined);
        },
    };
    return self;
}
