// @datadog/browser-rum/esm/domain/record/serialization/conversions/vNode.js
import { PlaybackState, NodeType } from '../../../../types';
export function createVNode(document, id, data) {
    let state = 'new';
    const assertConnectedCorrectly = () => {
        if (self.state !== 'connected') {
            throw new Error(`Expected node ${self.id} to be connected, not ${self.state}`);
        }
        for (const connection of ['firstChild', 'lastChild', 'previousSibling', 'nextSibling', 'parent']) {
            const connectedNode = self[connection];
            if (connectedNode && connectedNode.state !== 'connected') {
                throw new Error(`Expected node ${self.id}'s ${connection} to be connected, not ${connectedNode.state}`);
            }
        }
    };
    const self = {
        after(node) {
            var _a;
            node.parent = self.parent;
            node.previousSibling = self;
            if (self.nextSibling) {
                self.nextSibling.previousSibling = node;
            }
            node.nextSibling = self.nextSibling;
            self.nextSibling = node;
            if (((_a = self.parent) === null || _a === void 0 ? void 0 : _a.lastChild) === self) {
                self.parent.lastChild = node;
            }
            document.onNodeConnected(node, self.parent);
            assertConnectedCorrectly();
        },
        appendChild(node) {
            if (self.lastChild) {
                self.lastChild.after(node);
                return;
            }
            node.parent = self;
            self.firstChild = node;
            self.lastChild = node;
            document.onNodeConnected(node, self);
            assertConnectedCorrectly();
        },
        before(node) {
            var _a;
            node.parent = self.parent;
            node.nextSibling = self;
            if (self.previousSibling) {
                self.previousSibling.nextSibling = node;
            }
            node.previousSibling = self.previousSibling;
            self.previousSibling = node;
            if (((_a = self.parent) === null || _a === void 0 ? void 0 : _a.firstChild) === self) {
                self.parent.firstChild = node;
            }
            document.onNodeConnected(node, self.parent);
            assertConnectedCorrectly();
        },
        remove() {
            var _a, _b;
            if (self.state === 'disconnected') {
                return; // This is a redundant remove.
            }
            if (((_a = self.parent) === null || _a === void 0 ? void 0 : _a.firstChild) === self) {
                self.parent.firstChild = self.nextSibling;
            }
            if (((_b = self.parent) === null || _b === void 0 ? void 0 : _b.lastChild) === self) {
                self.parent.lastChild = self.previousSibling;
            }
            if (self.previousSibling) {
                self.previousSibling.nextSibling = self.nextSibling;
            }
            if (self.nextSibling) {
                self.nextSibling.previousSibling = self.previousSibling;
            }
            const parent = self.parent;
            self.parent = undefined;
            self.previousSibling = undefined;
            self.nextSibling = undefined;
            document.onNodeDisconnected(self, parent);
        },
        setAttachedStyleSheets(sheets) {
            assertConnectedCorrectly();
            switch (data.kind) {
                case '#document':
                case '#document-fragment':
                case '#element':
                case '#shadow-root':
                    data.attachedStyleSheets = sheets;
                    break;
                default:
                    throw new Error(`Cannot attach stylesheets to ${data.kind} node ${id}`);
            }
        },
        setAttribute(name, value) {
            assertConnectedCorrectly();
            if (data.kind !== '#element') {
                throw new Error(`Cannot set attribute ${name} on ${data.kind} node ${id}`);
            }
            if (value === null) {
                delete data.attributes[name];
            }
            else {
                data.attributes[name] = value;
            }
            document.onAttributeChanged(self, name);
        },
        setPlaybackState(state) {
            assertConnectedCorrectly();
            if (data.kind !== '#element') {
                throw new Error(`Cannot set media playback state of ${data.kind} node ${id}`);
            }
            data.playbackState = state;
        },
        setScrollPosition(left, top) {
            assertConnectedCorrectly();
            if (data.kind !== '#document' && data.kind !== '#element') {
                throw new Error(`Cannot set scroll position on ${data.kind} node ${id}`);
            }
            // Treat zero coordinates as 'not scrolled' unless this element has scrolled in the past.
            if (left !== 0 || data.scrollLeft !== undefined) {
                data.scrollLeft = left;
            }
            if (top !== 0 || data.scrollTop !== undefined) {
                data.scrollTop = top;
            }
        },
        setSize(width, height) {
            assertConnectedCorrectly();
            if (data.kind !== '#element') {
                throw new Error(`Cannot set size of ${data.kind} node ${id}`);
            }
            data.width = width;
            data.height = height;
        },
        setTextContent(value) {
            assertConnectedCorrectly();
            if (data.kind !== '#text') {
                throw new Error(`Cannot set text on ${data.kind} node ${id}`);
            }
            data.textContent = value;
            document.onTextChanged(self);
        },
        forEachAddedNodeRoot(nodeAdds, action) {
            if (nodeAdds.has(id)) {
                // This is the root of a newly-added subtree.
                action(self);
                return;
            }
            // This is an existing node, but there may be new nodes among our descendants. Visit
            // children in reverse order to match the ordering that the V1 serialization
            // algorithm would use.
            for (let child = self.lastChild; child; child = child.previousSibling) {
                child.forEachAddedNodeRoot(nodeAdds, action);
            }
        },
        forSelfAndEachDescendant(action) {
            action(self);
            for (let child = self.firstChild; child; child = child.nextSibling) {
                child.forSelfAndEachDescendant(action);
            }
        },
        mapChildren(action) {
            const results = [];
            for (let child = self.firstChild; child; child = child.nextSibling) {
                results.push(action(child));
            }
            return results;
        },
        render(options) {
            var _a, _b, _c, _d, _e;
            assertConnectedCorrectly();
            const id = (_b = (_a = options.nodeIdRemapper) === null || _a === void 0 ? void 0 : _a.remap(self.id)) !== null && _b !== void 0 ? _b : self.id;
            switch (data.kind) {
                case '#cdata-section':
                    return {
                        type: NodeType.CDATA,
                        id,
                        textContent: '',
                    };
                case '#doctype':
                    return {
                        type: NodeType.DocumentType,
                        id,
                        name: data.name,
                        publicId: data.publicId,
                        systemId: data.systemId,
                    };
                case '#document':
                    return {
                        type: NodeType.Document,
                        id,
                        childNodes: self.mapChildren((node) => node.render(options)),
                        adoptedStyleSheets: (_c = data.attachedStyleSheets) === null || _c === void 0 ? void 0 : _c.map((sheet) => sheet.renderAsAdoptedStyleSheet()),
                    };
                case '#document-fragment':
                    return {
                        type: NodeType.DocumentFragment,
                        id,
                        childNodes: self.mapChildren((node) => node.render(options)),
                        adoptedStyleSheets: (_d = data.attachedStyleSheets) === null || _d === void 0 ? void 0 : _d.map((sheet) => sheet.renderAsAdoptedStyleSheet()),
                        isShadowRoot: false,
                    };
                case '#element': {
                    const attributes = {};
                    // Add size-related virtual attributes before the real DOM attributes, to match
                    // the ordering used in the V1 format.
                    if (data.width !== undefined && data.height !== undefined) {
                        attributes.rr_width = `${data.width}px`;
                        attributes.rr_height = `${data.height}px`;
                    }
                    // Add DOM attributes.
                    Object.assign(attributes, data.attributes);
                    // Add other virtual attributes after the real DOM attributes, to match the
                    // ordering used in the V1 format.
                    if (data.attachedStyleSheets !== undefined) {
                        attributes._cssText = data.attachedStyleSheets.map((sheet) => sheet.renderAsCssText()).join('');
                    }
                    if (data.playbackState !== undefined) {
                        attributes.rr_mediaState = data.playbackState === PlaybackState.Paused ? 'paused' : 'played';
                    }
                    if (data.scrollLeft) {
                        attributes.rr_scrollLeft = data.scrollLeft;
                    }
                    if (data.scrollTop) {
                        attributes.rr_scrollTop = data.scrollTop;
                    }
                    return {
                        type: NodeType.Element,
                        id,
                        tagName: data.tag,
                        attributes,
                        childNodes: self.mapChildren((node) => node.render(options)),
                        isSVG: data.isSVG === true ? true : undefined,
                    };
                }
                case '#shadow-root':
                    return {
                        type: NodeType.DocumentFragment,
                        id,
                        childNodes: self.mapChildren((node) => node.render(options)),
                        adoptedStyleSheets: (_e = data.attachedStyleSheets) === null || _e === void 0 ? void 0 : _e.map((sheet) => sheet.renderAsAdoptedStyleSheet()),
                        isShadowRoot: true,
                    };
                case '#text':
                    return {
                        type: NodeType.Text,
                        id,
                        textContent: data.textContent,
                    };
                default:
                    data;
                    throw new Error(`Rendering not implemented for ${self.data.kind} node ${id}`);
            }
        },
        get data() {
            return data;
        },
        get ownerDocument() {
            return document;
        },
        get id() {
            return id;
        },
        get state() {
            return state;
        },
        set state(value) {
            if ((state === 'new' && value !== 'connected') ||
                (state === 'connected' && value !== 'disconnected') ||
                state === 'disconnected') {
                throw new Error(`Invalid state transition from ${state} to ${value}`);
            }
            state = value;
        },
        firstChild: undefined,
        lastChild: undefined,
        previousSibling: undefined,
        nextSibling: undefined,
        parent: undefined,
    };
    return self;
}
