// @datadog/browser-rum/cjs/domain/record/serialization/conversions/changeDecoder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createChangeDecoder = createChangeDecoder;
const types_1 = require("../../../../types");
const stringTable_1 = require("./stringTable");
function createChangeDecoder() {
    const self = {
        decode(record) {
            return decodeChangeRecord(record, self.stringTable);
        },
        stringTable: (0, stringTable_1.createStringTable)(),
    };
    return self;
}
function decodeChangeRecord(record, stringTable) {
    const decodedData = [];
    for (const change of record.data) {
        switch (change[0]) {
            case types_1.ChangeType.AddString:
                // Update the string table.
                for (let i = 1; i < change.length; i++) {
                    stringTable.add(change[i]);
                }
                // Deliberately don't include this change in the decoded record.
                break;
            case types_1.ChangeType.AddNode: {
                const decoded = [types_1.ChangeType.AddNode];
                for (let i = 1; i < change.length; i++) {
                    decoded.push(decodeAddNodeChange(change[i], stringTable));
                }
                decodedData.push(decoded);
                break;
            }
            case types_1.ChangeType.RemoveNode:
                decodedData.push(change);
                break;
            case types_1.ChangeType.Attribute: {
                const decoded = [types_1.ChangeType.Attribute];
                for (let i = 1; i < change.length; i++) {
                    decoded.push(decodeAttributeChange(change[i], stringTable));
                }
                decodedData.push(decoded);
                break;
            }
            case types_1.ChangeType.Text: {
                const decoded = [types_1.ChangeType.Text];
                for (let i = 1; i < change.length; i++) {
                    decoded.push(decodeTextChange(change[i], stringTable));
                }
                decodedData.push(decoded);
                break;
            }
            case types_1.ChangeType.Size:
            case types_1.ChangeType.ScrollPosition:
            case types_1.ChangeType.AttachedStyleSheets:
            case types_1.ChangeType.MediaPlaybackState:
            case types_1.ChangeType.VisualViewport:
                decodedData.push(change);
                break;
            case types_1.ChangeType.AddStyleSheet: {
                const decoded = [types_1.ChangeType.AddStyleSheet];
                for (let i = 1; i < change.length; i++) {
                    decoded.push(decodeAddStyleSheetChange(change[i], stringTable));
                }
                decodedData.push(decoded);
                break;
            }
            default:
                change;
                throw new Error(`Unsupported ChangeType: ${change[0]}`);
        }
    }
    return { ...record, data: decodedData };
}
function decodeAddNodeChange(change, stringTable) {
    const insertionPoint = change[0];
    const nodeName = stringTable.decode(change[1]);
    switch (nodeName) {
        case '#cdata-section':
        case '#document':
        case '#document-fragment':
        case '#shadow-root':
            return [insertionPoint, nodeName];
        case '#doctype': {
            const [, , name, publicId, systemId] = change;
            return [
                insertionPoint,
                '#doctype',
                stringTable.decode(name),
                stringTable.decode(publicId),
                stringTable.decode(systemId),
            ];
        }
        case '#text': {
            const [, , textContent] = change;
            return [insertionPoint, '#text', stringTable.decode(textContent)];
        }
        default: {
            const decodedChange = [insertionPoint, nodeName];
            const [, , ...attrs] = change;
            for (const [name, value] of attrs) {
                decodedChange.push([stringTable.decode(name), stringTable.decode(value)]);
            }
            return decodedChange;
        }
    }
}
function decodeAttributeChange(change, stringTable) {
    const [nodeId, ...mutations] = change;
    const decodedMutations = mutations.map((mutation) => {
        if (mutation.length === 1) {
            return [stringTable.decode(mutation[0])];
        }
        return [stringTable.decode(mutation[0]), stringTable.decode(mutation[1])];
    });
    const decodedChange = [nodeId];
    decodedChange.push(...decodedMutations);
    return decodedChange;
}
function decodeTextChange(change, stringTable) {
    return [change[0], stringTable.decode(change[1])];
}
function decodeAddStyleSheetChange(change, stringTable) {
    const rules = change[0];
    const decodedRules = Array.isArray(rules)
        ? rules.map((rule) => stringTable.decode(rule))
        : stringTable.decode(rules);
    if (change.length === 1) {
        return [decodedRules];
    }
    const decodedMediaList = change[1].map((item) => stringTable.decode(item));
    if (change.length === 2) {
        return [decodedRules, decodedMediaList];
    }
    return [decodedRules, decodedMediaList, change[2]];
}
