// @datadog/browser-rum/cjs/domain/record/serialization/conversions/changeConverter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createChangeConverter = createChangeConverter;
const types_1 = require("../../../../types");
const stringTable_1 = require("./stringTable");
const vDocument_1 = require("./vDocument");
function createChangeConverter() {
    const self = {
        convert(record, options = {}) {
            applyChangeToVDOM(record, self.document, self.stringTable);
            return self.document.render({ timestamp: record.timestamp, ...options });
        },
        document: (0, vDocument_1.createVDocument)(),
        stringTable: (0, stringTable_1.createStringTable)(),
    };
    return self;
}
function applyChangeToVDOM(record, document, stringTable) {
    document.mutations.clear();
    for (const change of record.data) {
        switch (change[0]) {
            case types_1.ChangeType.AddString: {
                for (let i = 1; i < change.length; i++) {
                    stringTable.add(change[i]);
                }
                break;
            }
            case types_1.ChangeType.AddNode: {
                for (let i = 1; i < change.length; i++) {
                    applyAddNodeChange(change[i], document, stringTable);
                }
                break;
            }
            case types_1.ChangeType.RemoveNode: {
                for (let i = 1; i < change.length; i++) {
                    applyRemoveNodeChange(change[i], document);
                }
                break;
            }
            case types_1.ChangeType.Attribute: {
                for (let i = 1; i < change.length; i++) {
                    applyAttributeChange(change[i], document, stringTable);
                }
                break;
            }
            case types_1.ChangeType.Text: {
                for (let i = 1; i < change.length; i++) {
                    applyTextChange(change[i], document, stringTable);
                }
                break;
            }
            case types_1.ChangeType.ScrollPosition: {
                for (let i = 1; i < change.length; i++) {
                    applyScrollPositionChange(change[i], document);
                }
                break;
            }
            case types_1.ChangeType.Size: {
                for (let i = 1; i < change.length; i++) {
                    applySizeChange(change[i], document);
                }
                break;
            }
            case types_1.ChangeType.AddStyleSheet: {
                for (let i = 1; i < change.length; i++) {
                    applyAddStyleSheetChange(change[i], document, stringTable);
                }
                break;
            }
            case types_1.ChangeType.AttachedStyleSheets: {
                for (let i = 1; i < change.length; i++) {
                    applyAttachedStyleSheetsChange(change[i], document);
                }
                break;
            }
            case types_1.ChangeType.MediaPlaybackState: {
                for (let i = 1; i < change.length; i++) {
                    applyMediaPlaybackStateChange(change[i], document);
                }
                break;
            }
            case types_1.ChangeType.VisualViewport:
                // Not yet supported.
                break;
            default:
                change;
                throw new Error(`Unsupported BrowserChangeType: ${change[0]}`);
        }
    }
}
function applyAddNodeChange(addedNode, document, stringTable) {
    const nodeName = stringTable.decode(addedNode[1]);
    let node;
    switch (nodeName) {
        case '#cdata-section':
            node = document.createNode({ kind: '#cdata-section' });
            break;
        case '#doctype': {
            const [, , name, publicId, systemId] = addedNode;
            node = document.createNode({
                kind: '#doctype',
                name: stringTable.decode(name),
                publicId: stringTable.decode(publicId),
                systemId: stringTable.decode(systemId),
            });
            break;
        }
        case '#document':
            node = document.createNode({ kind: '#document' });
            break;
        case '#document-fragment':
            node = document.createNode({ kind: '#document-fragment' });
            break;
        case '#shadow-root':
            node = document.createNode({ kind: '#shadow-root' });
            break;
        case '#text': {
            const [, , textContent] = addedNode;
            node = document.createNode({
                kind: '#text',
                textContent: stringTable.decode(textContent),
            });
            break;
        }
        default: {
            let tagName;
            let isSVG = false;
            if (nodeName.startsWith('svg>')) {
                tagName = nodeName.substring(4).toLowerCase();
                isSVG = true;
            }
            else {
                tagName = nodeName.toLowerCase();
            }
            const [, , ...attributeAssignments] = addedNode;
            const attributes = {};
            for (const [name, value] of attributeAssignments) {
                attributes[stringTable.decode(name)] = stringTable.decode(value);
            }
            node = document.createNode({ kind: '#element', tag: tagName, attributes, isSVG });
            break;
        }
    }
    const insertionPoint = addedNode[0];
    if (insertionPoint === null) {
        document.root = node;
    }
    else if (insertionPoint === 0) {
        const previousSiblingId = (node.id - 1);
        const previousSibling = document.getNodeById(previousSiblingId);
        previousSibling.after(node);
    }
    else if (insertionPoint > 0) {
        const parentId = (node.id - insertionPoint);
        const parent = document.getNodeById(parentId);
        parent.appendChild(node);
    }
    else {
        const nextSiblingId = (node.id + insertionPoint);
        const nextSibling = document.getNodeById(nextSiblingId);
        nextSibling.before(node);
    }
}
function applyAddStyleSheetChange(change, document, stringTable) {
    const [encodedRules, encodedMediaList = [], disabled = false] = change;
    const rules = Array.isArray(encodedRules)
        ? encodedRules.map((rule) => stringTable.decode(rule))
        : stringTable.decode(encodedRules);
    const mediaList = encodedMediaList.map((item) => stringTable.decode(item));
    document.createStyleSheet({ rules, mediaList, disabled });
}
function applyAttachedStyleSheetsChange(change, document) {
    const [nodeId, ...sheetIds] = change;
    const node = document.getNodeById(nodeId);
    node.setAttachedStyleSheets(sheetIds.map((sheetId) => document.getStyleSheetById(sheetId)));
}
function applyAttributeChange(change, document, stringTable) {
    const [nodeId, ...attributeMutations] = change;
    const node = document.getNodeById(nodeId);
    for (const [nameOrId, valueOrId = null] of attributeMutations) {
        const name = stringTable.decode(nameOrId);
        if (valueOrId === null) {
            node.setAttribute(name, null);
        }
        else {
            const value = stringTable.decode(valueOrId);
            node.setAttribute(name, value);
        }
    }
}
function applyMediaPlaybackStateChange(change, document) {
    const [nodeId, playbackState] = change;
    const node = document.getNodeById(nodeId);
    node.setPlaybackState(playbackState);
}
function applyRemoveNodeChange(change, document) {
    const nodeId = change;
    const node = document.getNodeById(nodeId);
    node.remove();
}
function applyScrollPositionChange(change, document) {
    const [nodeId, left, top] = change;
    const node = document.getNodeById(nodeId);
    node.setScrollPosition(left, top);
}
function applySizeChange(change, document) {
    const [nodeId, width, height] = change;
    const node = document.getNodeById(nodeId);
    node.setSize(width, height);
}
function applyTextChange(change, document, stringTable) {
    const [nodeId, textContent] = change;
    const node = document.getNodeById(nodeId);
    node.setTextContent(stringTable.decode(textContent));
}
