// @datadog/browser-rum/cjs/domain/record/serialization/conversions/nodeIdRemapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createIdentityNodeIdRemapper = createIdentityNodeIdRemapper;
exports.createCopyingNodeIdRemapper = createCopyingNodeIdRemapper;
const itemIds_1 = require("../../itemIds");
function createIdentityNodeIdRemapper() {
    return {
        remap(inputNodeId) {
            return inputNodeId;
        },
    };
}
function createCopyingNodeIdRemapper() {
    const { nodeIds: inputNodeIds, nodesByNodeId: inputNodesByNodeId } = createTrackedNodeIds();
    const { nodeIds: outputNodeIds, nodeIdsByNode: outputNodeIdsByNode } = createTrackedNodeIds();
    const self = {
        remap(inputNodeId) {
            var _a;
            const node = inputNodesByNodeId.get(inputNodeId);
            if (!node) {
                throw new Error(`Input node id ${inputNodeId} not found`);
            }
            return (_a = outputNodeIdsByNode.get(node)) !== null && _a !== void 0 ? _a : outputNodeIds.getOrInsert(node);
        },
        inputNodeIds,
        outputNodeIds,
    };
    return self;
}
function createTrackedNodeIds() {
    const wrappedNodeIds = (0, itemIds_1.createNodeIds)();
    const nodeIdsByNode = new Map();
    const nodesByNodeId = new Map();
    const nodeIds = {
        clear: wrappedNodeIds.clear,
        delete: wrappedNodeIds.delete,
        get: wrappedNodeIds.get,
        getOrInsert(node) {
            const nodeId = wrappedNodeIds.getOrInsert(node);
            nodeIdsByNode.set(node, nodeId);
            nodesByNodeId.set(nodeId, node);
            return nodeId;
        },
        get nextId() {
            return wrappedNodeIds.nextId;
        },
        get size() {
            return wrappedNodeIds.size;
        },
    };
    return { nodeIds, nodeIdsByNode, nodesByNodeId };
}
