// @datadog/browser-rum-core/cjs/domain/resource/requestRegistry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MAX_REQUESTS = void 0;
exports.createRequestRegistry = createRequestRegistry;
const browser_core_1 = require("@datadog/browser-core");
// Maximum number of requests to keep in the registry. Requests should be removed quite quickly in
// general, this is just a safety limit to avoid memory leaks in case of a bug.
exports.MAX_REQUESTS = 1000;
function createRequestRegistry(lifeCycle) {
    const requests = new Set();
    let tooManyRequestsReported = false;
    const subscription = lifeCycle.subscribe(8 /* LifeCycleEventType.REQUEST_COMPLETED */, (request) => {
        requests.add(request);
        if (requests.size > exports.MAX_REQUESTS) {
            const oldestRequest = requests.values().next().value;
            if (!tooManyRequestsReported) {
                tooManyRequestsReported = true;
                let debugContext;
                if ((0, browser_core_1.isExperimentalFeatureEnabled)(browser_core_1.ExperimentalFeature.TOO_MANY_REQUESTS_INVESTIGATION)) {
                    let abortedCount = 0;
                    let abortedOnStartCount = 0;
                    let xhrCount = 0;
                    let withoutMatchingEntryCount = 0;
                    for (const r of requests) {
                        if (r.isAborted) {
                            abortedCount++;
                        }
                        if (r.isAbortedOnStart) {
                            abortedOnStartCount++;
                        }
                        if (r.type === browser_core_1.RequestType.XHR) {
                            xhrCount++;
                        }
                        const entries = performance.getEntriesByName(r.url, 'resource');
                        const hasMatchingEntry = entries.some((e) => e.startTime >= r.startClocks.relative);
                        if (!hasMatchingEntry) {
                            withoutMatchingEntryCount++;
                        }
                    }
                    const oldestRequestAge = (0, browser_core_1.timeStampNow)() - oldestRequest.startClocks.timeStamp;
                    debugContext = {
                        abortedCount,
                        abortedOnStartCount,
                        xhrCount,
                        fetchCount: requests.size - xhrCount,
                        oldestRequestAge,
                        oldestRequestEndAge: oldestRequestAge - oldestRequest.duration,
                        withoutMatchingEntryCount,
                    };
                }
                // monitor-until: 2026-06-01, after early request collection is the default in v7
                (0, browser_core_1.addTelemetryDebug)('Too many requests', debugContext);
            }
            requests.delete(oldestRequest);
        }
    });
    return {
        getMatchingRequest(entry) {
            // Returns the closest request object that happened before the entry
            let minTimeDifference = Infinity;
            let closestRequest;
            for (const request of requests) {
                const timeDifference = entry.startTime - request.startClocks.relative;
                if (0 <= timeDifference && timeDifference < minTimeDifference && request.url === entry.name) {
                    minTimeDifference = Math.abs(timeDifference);
                    closestRequest = request;
                }
            }
            if (closestRequest) {
                requests.delete(closestRequest);
            }
            return closestRequest;
        },
        stop() {
            subscription.unsubscribe();
        },
    };
}
